### PR TITLE
ASoC: SOF: ipc: add retry logic for IPC timeouts

### PR DESCRIPTION
**Pull request by kv2019i**
One idea how to keep using MSI interrupts. Based on debugging state of the interrupt tree status registers, the hardware state seems legit in cases where we have IPC timeouts, but the MSI generation somehow fails. Here's a simple recovery option (idea came up today).

In case IPC times out, try to recover by directly calling
the IRQ handler thread. This can help to recover in cases
where the IRQ was lost for some reason, but status registers
have valid content.

### CODE REVIEW

**Comment by kv2019i** [comment]
> I am not a big fan of this... What are the odds it'll be accepted upstream?

@plbossart Yeah, we won't be finding many fans, that's for sure. For upstream, not sure, there are similar logic in legacy HDA driver as well -- so if this is needed then it's needed, hard to argue with that. Few gems from existing upstream HDA code:

hda_controller.c:azx_rirb_get_response()
```
»       »       if (hbus->needs_damn_long_delay || loopcounter > 3000)
»       »       »       msleep(2); /* temporary workaround */
»       »       else {
»
```
...

```
»       »       if (chip->ops->disable_msi_reset_irq &&
»       »           chip->ops->disable_msi_reset_irq(chip) < 0)
»       »       »       return -EIO;
»       »       goto again;
```

But yeah, priority/preference is to figure out why we are losing MSI irqs.

File: sound/soc/sof/ipc.c [hunk 1]
```diff
@@ -207,11 +207,23 @@ static int tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
...
+	while (!ret && retries++ < 2) {
+		ret = wait_event_timeout(msg->waitq, msg->ipc_complete,
+					 msecs_to_jiffies(IPC_TIMEOUT_MS));
+		if (ret == 0) {
+			dev_warn(sdev->dev, "ipc timeout, polling irq thread\n");
+			if (sof_ops(sdev) && sof_ops(sdev)->irq_thread) {
+				disable_irq(sdev->ipc_irq);
+				ret = sof_ops(sdev)->irq_thread(sdev->ipc_irq,
+								sdev);
+				enable_irq(sdev->ipc_irq);
```
Comment (dbaluta) [inline sound/soc/sof/ipc.c:223]: It looks super strange to me for IPC framework directly call irq_thread function. For i.MX8 the IPC is wrapped around drivers/mailbox infrastructure and the client of the mailbox doesn't directly have access to IRQ handler. It only provides a callback that IRQ handler will call. :)

Not sure if we can somehow abstract the IPC communication using the mailbox API. One thing is for sure IPC looks hard to get right and it seems that everyone has its own framework :)

**Comment by plbossart** [comment]
@kv2019i this looks like an idea that will not be pursued further, time to close and move on?

File: sound/soc/sof/ipc.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/ipc.c:223]: @dbaluta Thanks -- you are right this makes  assumptions about irq handling that are not part of the current ipc device abstractions. Fortunately it seems we can close this pull req anyways as another better solution was found. Probably if something similar is needed in the future, it needs to be done in the hw dependent backend, and generic ipc code jsut gets the snd_sof_ipc_reply().

@plbossart Ack, not needed anymore, I'll close this.

**Comment by kv2019i** [comment]
PR1013 will be used instead, closing this one.

**Pull request closed without merging**